neural_lam/weather_dataset.py

```diff
@@ -1,262 +1,373 @@
+                da_forcing_windowed.values, dtype=torch.float32
+            )
+
+        # init_states: (2, N_grid, d_features)
+        # target_states: (ar_steps, N_grid, d_features)
+        # forcing: (ar_steps, N_grid, d_windowed_forcing)
+        # batch_times: (ar_steps,)
+
+        return init_states, target_states, forcing, batch_times
+
+    def __iter__(self):
+        """
+        Convenience method to iterate over the dataset.
+
+        This isn't used by pytorch DataLoader which itself implements an
+        iterator that uses Dataset.__getitem__ and Dataset.__len__.
+
+        """
+        for i in range(len(self)):
+            yield self[i]
+
+
+class WeatherDataModule(pl.LightningDataModule):
+    """DataModule for weather data."""
+
+    def __init__(
+        self,
+        datastore: BaseDatastore,
+        ar_steps_train=3,
+        ar_steps_eval=25,
+        standardize=True,
+        forcing_window_size=3,
+        batch_size=4,
+        num_workers=16,
+    ):
+        super().__init__()
+        self._datastore = datastore
+        self.forcing_window_size = forcing_window_size
+        self.ar_steps_train = ar_steps_train
+        self.ar_steps_eval = ar_steps_eval
+        self.standardize = standardize
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.train_dataset = None
+        self.val_dataset = None
+        self.test_dataset = None
+        if num_workers > 0:
+            # default to spawn for now, as the default on linux "fork" hangs
+            # when using dask (which the npyfiles datastore uses)
+            self.multiprocessing_context = "spawn"
+        else:
+            self.multiprocessing_context = None
+
+    def setup(self, stage=None):
+        if stage == "fit" or stage is None:
+            self.train_dataset = WeatherDataset(
+                datastore=self._datastore,
+                split="train",
+                ar_steps=self.ar_steps_train,
+                standardize=self.standardize,
+                forcing_window_size=self.forcing_window_size,
+            )
+            self.val_dataset = WeatherDataset(
+                datastore=self._datastore,
+                split="val",
+                ar_steps=self.ar_steps_eval,
+                standardize=self.standardize,
+                forcing_window_size=self.forcing_window_size,
+            )
+
+        if stage == "test" or stage is None:
+            self.test_dataset = WeatherDataset(
+                datastore=self._datastore,
+                split="test",
+                ar_steps=self.ar_steps_eval,
+                standardize=self.standardize,
+                forcing_window_size=self.forcing_window_size,
+            )
+
+    def train_dataloader(self):
+        """Load train dataset."""
+        return torch.utils.data.DataLoader(
+            self.train_dataset,
+            batch_size=self.batch_size,
+            num_workers=self.num_workers,
+            shuffle=False,
```

Comment:
We want to shuffle the training dataset. Since subsequent sample indices will correspond to subsequent initialization times not shuffling would result in large correlations between the samples in a batch.

Reply:
```suggestion
            shuffle=True,
```

Reply:
How does this relate to `use_distributed_sampler` proposed in https://github.com/mllam/neural-lam/pull/26? I am uncertain how these two arguments work in together.

Reply:
> We want to shuffle the training dataset. Since subsequent sample indices will correspond to subsequent initialization times not shuffling would result in large correlations between the samples in a batch.

That makes sense to me, but I don't think we do that currently: https://github.com/mllam/neural-lam/blob/main/neural_lam/train_model.py#L250 - do we need to introduce shuffling here because of how I have structure the datastores code @joeloskarsson?

Reply:
Note that this is specifically the *training set* I mention. What you link to is the validation set (which I don't think we should shuffle). We do currently shuffle the training set: https://github.com/mllam/neural-lam/blob/68399f7e0a3964bc73ec798b8b3c08cd761a3eb1/neural_lam/train_model.py#L226-L238 So I don't propose a  change based on the new datastores, just want to keep this consistent with the current setup, which I think follows good practice.

Reply:
Regarding the `use_distributed_sampler` in #26 : I don't believe that the shuffling should create an issues for that. But that is something that has to be tested as part of #26. Since we do shuffle the training set now, and I think #26 works, there is no reason to believe this introduces any new problems.